### PR TITLE
Reactive Repository Refactoring

### PR DESCRIPTION
**Pull request by ErikHellman**

### CODE REVIEW

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/TasksRepository.java [hunk 1]
```diff
@@ -234,78 +146,25 @@ public void clearCompletedTasks() {
...
-                    return task;
-                });
+        return mTasksLocalDataSource.getTask(taskId);
     }
 
     @Override
     public void refreshTasks() {
-        mCacheIsDirty = true;
+        mTasksRemoteDataSource.getTasks()
+                .subscribeOn(Schedulers.io())
```
Comment (florina-muntenescu) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/TasksRepository.java]: In the rxjava samples, we've been using a `BaseSchedulerProvider` for scheduling, and then that class gets to be mocked in tests. Should we have a consistent approach?

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/local/TasksLocalDataSource.java [hunk 2]
```diff
@@ -118,12 +119,28 @@ public static void destroyInstance() {
...
     public void saveTask(@NonNull Task task) {
         checkNotNull(task);
+        ContentValues values = toContentValues(task);
+        mDatabaseHelper.insert(TaskEntry.TABLE_NAME, values, SQLiteDatabase.CONFLICT_REPLACE);
+    }
+
+    @Override
+    public Completable saveTasks(@NonNull List<Task> tasks) {
+        return Observable.using(mDatabaseHelper::newTransaction,
+                transaction -> Observable.from(tasks)
```
Comment (florina-muntenescu) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/local/TasksLocalDataSource.java]: could you extract `Observable.from .... .toList()` to another method, to make things a bit more readable?

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/remote/TasksRemoteDataSource.java [hunk 3]
```diff
@@ -87,15 +88,21 @@ public void saveTask(@NonNull Task task) {
     }
 
     @Override
+    public Completable saveTasks(@NonNull List<Task> tasks) {
+        return Observable.from(tasks).doOnNext(this::saveTask).toCompletable();
```
Comment (florina-muntenescu) [inline todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/remote/TasksRemoteDataSource.java]: could you put `doOnNext` and `toCompletable` on different lines

File: todoapp/app/src/mock/java/com/example/android/architecture/blueprints/todoapp/data/FakeTasksRemoteDataSource.java [hunk 4]
```diff
@@ -52,44 +56,64 @@ public static FakeTasksRemoteDataSource getInstance() {
     @Override
     public Observable<List<Task>> getTasks() {
         Collection<Task> values = TASKS_SERVICE_DATA.values();
-        return Observable.from(values).toList();
+        return Observable.from(values).toList().repeatWhen(new Func1<Observable<? extends Void>, Observable<Boolean>>() {
```
Comment (florina-muntenescu) [inline todoapp/app/src/mock/java/com/example/android/architecture/blueprints/todoapp/data/FakeTasksRemoteDataSource.java]: Isn't `Observable.from(values).toList()` == `Observable.just(values)` ? 
Also, can we use lambdas? 
and put `repeatWhen` on a different line.

File: todoapp/app/src/mock/java/com/example/android/architecture/blueprints/todoapp/data/FakeTasksRemoteDataSource.java [hunk 5]
```diff
@@ -52,44 +56,64 @@ public static FakeTasksRemoteDataSource getInstance() {
...
+                return repeatWhen;
+            }
+        });
     }
 
     @Override
     public Observable<Task> getTask(@NonNull String taskId) {
         Task task = TASKS_SERVICE_DATA.get(taskId);
-        return Observable.just(task);
+        return Observable.just(task).repeatWhen(new Func1<Observable<? extends Void>, Observable<Boolean>>() {
```
Comment (florina-muntenescu) [inline todoapp/app/src/mock/java/com/example/android/architecture/blueprints/todoapp/data/FakeTasksRemoteDataSource.java]: `repeatWhen` on a different line and lambda

File: todoapp/app/src/mock/java/com/example/android/architecture/blueprints/todoapp/data/FakeTasksRemoteDataSource.java [hunk 6]
```diff
@@ -52,44 +56,64 @@ public static FakeTasksRemoteDataSource getInstance() {
...
 
     @Override
     public void saveTask(@NonNull Task task) {
         TASKS_SERVICE_DATA.put(task.getId(), task);
+        repeatWhen.onNext(true);
+    }
+
+    @Override
+    public Completable saveTasks(@NonNull List<Task> tasks) {
+        return Observable.from(tasks).doOnNext(this::saveTask).toCompletable();
```
Comment (florina-muntenescu) [inline todoapp/app/src/mock/java/com/example/android/architecture/blueprints/todoapp/data/FakeTasksRemoteDataSource.java]: can we put every operator on a new line?

File: todoapp/app/src/test/java/com/example/android/architecture/blueprints/todoapp/data/source/TasksRepositoryTest.java [hunk 7]
```diff
@@ -162,7 +162,7 @@ public void saveTask_savesTaskToServiceAPI() {
         // Then the service API and persistent repository are called and the cache is updated
         verify(mTasksRemoteDataSource).saveTask(newTask);
         verify(mTasksLocalDataSource).saveTask(newTask);
-        assertThat(mTasksRepository.mCachedTasks.size(), is(1));
+        ;
```
Comment (florina-muntenescu) [inline todoapp/app/src/test/java/com/example/android/architecture/blueprints/todoapp/data/source/TasksRepositoryTest.java]: remove ";"

File: todoapp/app/src/mock/java/com/example/android/architecture/blueprints/todoapp/data/FakeTasksRemoteDataSource.java [hunk 8]
```diff
@@ -38,6 +41,7 @@
 
     private static final Map<String, Task> TASKS_SERVICE_DATA = new LinkedHashMap<>();
 
+    private PublishSubject<Boolean> repeatWhen = PublishSubject.create();
```
Comment (florina-muntenescu) [inline todoapp/app/src/mock/java/com/example/android/architecture/blueprints/todoapp/data/FakeTasksRemoteDataSource.java]: `mRepeatWhen`

Also, can you add a small comment on why we are using this?

File: todoapp/app/src/test/java/com/example/android/architecture/blueprints/todoapp/data/source/TasksRepositoryTest.java [hunk 9]
```diff
@@ -109,7 +109,7 @@ public void getTasks_repositoryCachesAfterFirstSubscription_whenTasksAvailableIn
         verify(mTasksRemoteDataSource).getTasks();
         verify(mTasksLocalDataSource).getTasks();
         //
```
Comment (florina-muntenescu) [inline todoapp/app/src/test/java/com/example/android/architecture/blueprints/todoapp/data/source/TasksRepositoryTest.java]: remove //

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/TasksRepository.java (hunk 1 above)
Comment (ErikHellman) [reply todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/TasksRepository.java]: Yes! Sounds lika a good approach.

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/local/TasksLocalDataSource.java (hunk 2 above)
Comment (ErikHellman) [reply todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/local/TasksLocalDataSource.java]: 👍

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/remote/TasksRemoteDataSource.java (hunk 3 above)
Comment (ErikHellman) [reply todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/data/source/remote/TasksRemoteDataSource.java]: 👍

File: todoapp/app/src/mock/java/com/example/android/architecture/blueprints/todoapp/data/FakeTasksRemoteDataSource.java (hunk 4 above)
Comment (ErikHellman) [reply todoapp/app/src/mock/java/com/example/android/architecture/blueprints/todoapp/data/FakeTasksRemoteDataSource.java]: Yes. My bad. :)

File: todoapp/app/src/mock/java/com/example/android/architecture/blueprints/todoapp/data/FakeTasksRemoteDataSource.java (hunk 5 above)
Comment (ErikHellman) [reply todoapp/app/src/mock/java/com/example/android/architecture/blueprints/todoapp/data/FakeTasksRemoteDataSource.java]: 👍

File: todoapp/app/src/mock/java/com/example/android/architecture/blueprints/todoapp/data/FakeTasksRemoteDataSource.java (hunk 6 above)
Comment (ErikHellman) [reply todoapp/app/src/mock/java/com/example/android/architecture/blueprints/todoapp/data/FakeTasksRemoteDataSource.java]: 👍

File: todoapp/app/src/test/java/com/example/android/architecture/blueprints/todoapp/data/source/TasksRepositoryTest.java (hunk 7 above)
Comment (ErikHellman) [reply todoapp/app/src/test/java/com/example/android/architecture/blueprints/todoapp/data/source/TasksRepositoryTest.java]: I was trying to make a smiley? 8)

File: todoapp/app/src/mock/java/com/example/android/architecture/blueprints/todoapp/data/FakeTasksRemoteDataSource.java (hunk 8 above)
Comment (ErikHellman) [reply todoapp/app/src/mock/java/com/example/android/architecture/blueprints/todoapp/data/FakeTasksRemoteDataSource.java]: 👍

**Pull request merged**